File: java/com/google/copybara/action/ActionContext.java
/*
 * Copyright (C) 2021 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.action;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.copybara.config.SkylarkUtil.convertFromNoneable;
import static com.google.copybara.exception.ValidationException.checkCondition;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.copybara.DestinationEffect;
import com.google.copybara.DestinationEffect.DestinationRef;
import com.google.copybara.DestinationEffect.OriginRef;
import com.google.copybara.SkylarkContext;
import com.google.copybara.config.SkylarkUtil;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.transform.SkylarkConsole;
import com.google.copybara.util.console.Console;

import net.starlark.java.annot.Param;
import net.starlark.java.annot.ParamType;
import net.starlark.java.annot.StarlarkMethod;
import net.starlark.java.eval.Dict;
import net.starlark.java.eval.EvalException;
import net.starlark.java.eval.NoneType;
import net.starlark.java.eval.Sequence;
import net.starlark.java.eval.StarlarkValue;

import java.util.ArrayList;
import java.util.List;

/**
 * A StarlarkContext for running Actions.
 */
public abstract class ActionContext<T extends SkylarkContext<T>> implements SkylarkContext<T>,
    StarlarkValue {

  protected final List<DestinationEffect> newDestinationEffects = new ArrayList<>();
  protected final Action action;
  protected final SkylarkConsole console;
  protected final ImmutableMap<String, String> labels;
  private final Dict<?, ?> params;
  private ActionResult actionResult;

  public ActionContext(Action action, SkylarkConsole console, ImmutableMap<String, String> labels,
      Dict<?, ?> params) {
    this.action = checkNotNull(action);
    this.console = checkNotNull(console);
    this.labels = checkNotNull(labels);
    this.params = checkNotNull(params);
  }

  @StarlarkMethod(
      name = "action_name",
      doc = "The name of the current action.",
      structField = true)
  public String getActionName() {
    return action.getName();
  }

  @StarlarkMethod(name = "console", doc = "Get an instance of the console to report errors or"
      + " warnings", structField = true)
  public Console getConsole() {
    return console;
  }

  @StarlarkMethod(
      name = "params",
      doc = "Parameters for the function if created with" + " core.action",
      structField = true)
  public Dict<?, ?> getParams() {
    return params;
  }

  @StarlarkMethod(name = "success", doc = "Returns a successful action result.")
  public ActionResult success() {
    return ActionResult.success();
  }

  @StarlarkMethod(
      name = "noop",
      doc = "Returns a no op action result with an optional message.",
      parameters = {
          @Param(
              name = "msg",
              allowedTypes = {
                  @ParamType(type = String.class),
                  @ParamType(type = NoneType.class),
              },
              doc = "The no op message",
              defaultValue = "None"),
      })
  public ActionResult noop(Object noopMsg) {
    return ActionResult.noop(convertFromNoneable(noopMsg, /*defaultMsg*/ null));
  }

  @StarlarkMethod(
      name = "error",
      doc = "Returns an error action result.",
      parameters = {
          @Param(name = "msg", doc = "The error message"),
      })
  public ActionResult error(String errorMsg) {
    return ActionResult.error(errorMsg);
  }

  /**
   * Return the new {@link DestinationEffect}s created by this context.
   */
  public ImmutableList<DestinationEffect> getNewDestinationEffects() {
    return ImmutableList.copyOf(newDestinationEffects);
  }

  @StarlarkMethod(
      name = "cli_labels",
      doc = "Access labels that a user passes through flag '--labels'. "
          + "For example: --labels=foo:value1,bar:value2. Then it can access in this way:"
          + "cli_labels['foo'].",
      structField = true)
  public Dict<String, String> getCliLabels() {
    return Dict.copyOf(null, labels);
  }

  @StarlarkMethod(
      name = "record_effect",
      doc = "Records an effect of the current action.",
      parameters = {
          @Param(name = "summary", doc = "The summary of this effect", named = true),
          @Param(
              name = "origin_refs",
              allowedTypes = {
                  @ParamType(type = Sequence.class, generic1 = OriginRef.class),
              },
              doc = "The origin refs",
              named = true),
          @Param(name = "destination_ref", doc = "The destination ref", named = true),
          @Param(
              name = "errors",
              allowedTypes = {@ParamType(type = Sequence.class, generic1 = String.class)},
              defaultValue = "[]",
              doc = "An optional list of errors",
              named = true),
          @Param(
              name = "type",
              doc =
                  "The type of migration effect:<br>"
                      + "<ul>"
                      + "<li><b>'CREATED'</b>: A new review or change was created.</li>"
                      + "<li><b>'UPDATED'</b>: An existing review or change was updated.</li>"
                      + "<li><b>'NOOP'</b>: The change was a noop.</li>"
                      + "<li><b>'NOOP_AGAINST_PENDING_CHANGE'</b>: The change was a noop, relative"
                      + "to an existing pending change.</li>"
                      + "<li><b>'INSUFFICIENT_APPROVALS'</b>: The effect couldn't happen because "
                      + "the change doesn't have enough approvals.</li>"
                      + "<li><b>'ERROR'</b>: A user attributable error happened that prevented "
                      + "the destination from creating/updating the change. "
                      + "<li><b>'STARTED'</b>: The initial effect of a migration that depends on a "
                      + "previous one. This allows to have 'dependant' migrations defined by users.\n"
                      + "An example of this: a workflow migrates code from a Gerrit review to a "
                      + "GitHub PR, and a feedback migration migrates the test results from a CI in "
                      + "GitHub back to the Gerrit change.\n"
                      + "This effect would be created on the former one.</li>"
                      + "</ul>",
              defaultValue = "\"UPDATED\"",
              named = true)
      })
  public void recordEffect(
      String summary,
      Sequence<?> originRefs, // <OriginRef>
      DestinationRef destinationRef,
      Sequence<?> errors, // <String>
      String typeStr)
      throws EvalException {
    DestinationEffect.Type type =
        SkylarkUtil.stringToEnum("type", typeStr, DestinationEffect.Type.class);
    newDestinationEffects.add(
        new DestinationEffect(
            type,
            summary,
            Sequence.cast(originRefs, OriginRef.class, "origin_refs"),
            destinationRef,
            Sequence.cast(errors, String.class, "errors")));
  }

  @Override
  public void onFinish(Object result, SkylarkContext<T> context) throws ValidationException {
    checkCondition(
        result != null,
        "Actions must return a result via built-in functions: success(), "
            + "error(), noop() return, but '%s' returned: None", action.getName());
    checkCondition(result instanceof ActionResult,
        "Actions must return a result via built-in functions: success(), "
            + "error(), noop() return, but '%s' returned: %s", action.getName(), result);
    this.actionResult = (ActionResult) result;
    switch (actionResult.getResult()) {
      case ERROR:
        console.errorFmt(
            "Action '%s' returned error: %s", action.getName(), actionResult.getMsg());
        break;
      case NO_OP:
        console.infoFmt(
            "Action '%s' returned noop: %s", action.getName(), actionResult.getMsg());
        break;
      case SUCCESS:
        console.infoFmt("Action '%s' returned success", action.getName());
        break;
    }
    // Populate effects registered in the action context. This is required because SkylarkAction
    // makes a copy of the context to inject the parameters, but that instance is not visible from
    // the caller
    this.newDestinationEffects.addAll(((ActionContext<T>) context).newDestinationEffects);
  }

  public ActionResult getActionResult() {
    return actionResult;
  }
}

File: src/Copybara.Core/Action/ActionContext.cs
/*
 * Copyright (C) 2021 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using Copybara.Common;
using Copybara.Config;
using Copybara.Effect;
using Copybara.Exceptions;
using Copybara.Revision;
using Copybara.Transform;
using Starlark.Annot;
using Starlark.Eval;
using ConsoleT = Copybara.Util.Console.Console;
using StarlarkRt = Starlark.Eval.Starlark;
using Sequence = Starlark.Eval.Sequence;

namespace Copybara.Action;

/// <summary>
/// A StarlarkContext for running Actions. Port of <c>com.google.copybara.action.ActionContext</c>.
/// </summary>
public abstract class ActionContext<T> : ISkylarkContext<T>, IStarlarkValue
    where T : ISkylarkContext<T>
{
    protected readonly List<DestinationEffect> NewDestinationEffects = new();
    private ActionResult? _actionResult;

    protected ActionContext(
        IAction action,
        SkylarkConsole console,
        IReadOnlyDictionary<string, string> labels,
        Dict @params)
    {
        Action = Preconditions.CheckNotNull(action);
        Console = Preconditions.CheckNotNull(console);
        Labels = Preconditions.CheckNotNull(labels);
        Params = Preconditions.CheckNotNull(@params);
    }

    /// <summary>The action this context runs.</summary>
    protected IAction Action { get; }

    /// <summary>The console used to report errors or warnings.</summary>
    protected SkylarkConsole Console { get; }

    /// <summary>The CLI labels passed to the migration.</summary>
    protected IReadOnlyDictionary<string, string> Labels { get; }

    private Dict Params { get; }

    [StarlarkMethod(
        "action_name",
        Doc = "The name of the current action.",
        StructField = true)]
    public string GetActionName() => Action.GetName();

    [StarlarkMethod(
        "console",
        Doc = "Get an instance of the console to report errors or warnings",
        StructField = true)]
    public ConsoleT GetConsole() => Console;

    [StarlarkMethod(
        "params",
        Doc = "Parameters for the function if created with core.action",
        StructField = true)]
    public Dict GetParams() => Params;

    [StarlarkMethod("success", Doc = "Returns a successful action result.")]
    public ActionResult Success() => ActionResult.SuccessResult();

    [StarlarkMethod(
        "noop",
        Doc = "Returns a no op action result with an optional message.")]
    public ActionResult Noop(
        [Param(
            Name = "msg",
            AllowedTypes = new[] { typeof(string), typeof(NoneType) },
            Doc = "The no op message",
            DefaultValue = "None")]
        object? noopMsg) =>
        ActionResult.NoopResult(SkylarkUtil.ConvertFromNoneable<string>(noopMsg, null));

    [StarlarkMethod(
        "error",
        Doc = "Returns an error action result.")]
    public ActionResult Error(
        [Param(Name = "msg", Doc = "The error message")] string errorMsg) =>
        ActionResult.ErrorResult(errorMsg);

    /// <summary>Return the new <see cref="DestinationEffect"/>s created by this context.</summary>
    public IReadOnlyList<DestinationEffect> GetNewDestinationEffects() =>
        NewDestinationEffects.ToImmutableArray();

    [StarlarkMethod(
        "cli_labels",
        Doc = "Access labels that a user passes through flag '--labels'. "
            + "For example: --labels=foo:value1,bar:value2. Then it can access in this way:"
            + "cli_labels['foo'].",
        StructField = true)]
    public Dict GetCliLabels() =>
        Dict.ImmutableCopyOf(
            Labels.Select(kv => new KeyValuePair<object?, object?>(kv.Key, kv.Value)));

    [StarlarkMethod(
        "record_effect",
        Doc = "Records an effect of the current action.")]
    public void RecordEffect(
        [Param(Name = "summary", Doc = "The summary of this effect", Named = true)]
        string summary,
        [Param(
            Name = "origin_refs",
            AllowedTypes = new[] { typeof(ISequence<object>) },
            Doc = "The origin refs",
            Named = true)]
        ISequence<object?> originRefs,
        [Param(Name = "destination_ref", Doc = "The destination ref", Named = true)]
        DestinationEffect.DestinationRef destinationRef,
        [Param(
            Name = "errors",
            AllowedTypes = new[] { typeof(ISequence<object>) },
            DefaultValue = "[]",
            Doc = "An optional list of errors",
            Named = true)]
        ISequence<object?> errors,
        [Param(
            Name = "type",
            Doc = "The type of migration effect (CREATED, UPDATED, NOOP,"
                + " NOOP_AGAINST_PENDING_CHANGE, INSUFFICIENT_APPROVALS, ERROR, STARTED).",
            DefaultValue = "\"UPDATED\"",
            Named = true)]
        string typeStr)
    {
        var type = SkylarkUtil.StringToEnum<DestinationEffect.EffectType>("type", typeStr);
        NewDestinationEffects.Add(
            new DestinationEffect(
                type,
                summary,
                Sequence.Cast<OriginRef>(originRefs, "origin_refs"),
                destinationRef,
                Sequence.Cast<string>(errors, "errors")));
    }

    public abstract T WithParams(Dict @params);

    public void OnFinish(object? result, object context)
    {
        ValidationException.CheckCondition(
            result != null,
            "Actions must return a result via built-in functions: success(), "
                + "error(), noop() return, but '{0}' returned: None", Action.GetName());
        ValidationException.CheckCondition(
            result is ActionResult,
            "Actions must return a result via built-in functions: success(), "
                + "error(), noop() return, but '{0}' returned: {1}", Action.GetName(), result!);
        _actionResult = (ActionResult)result!;
        switch (_actionResult.GetResult())
        {
            case ActionResult.Result.Error:
                Console.ErrorFmt(
                    "Action '{0}' returned error: {1}", Action.GetName(), _actionResult.GetMsg()!);
                break;
            case ActionResult.Result.NoOp:
                Console.InfoFmt(
                    "Action '{0}' returned noop: {1}", Action.GetName(), _actionResult.GetMsg()!);
                break;
            case ActionResult.Result.Success:
                Console.InfoFmt("Action '{0}' returned success", Action.GetName());
                break;
        }

        // Populate effects registered in the action context. This is required because StarlarkAction
        // makes a copy of the context to inject the parameters, but that instance is not visible from
        // the caller.
        if (context is ActionContext<T> other)
        {
            NewDestinationEffects.AddRange(other.NewDestinationEffects);
        }
    }

    public ActionResult? GetActionResult() => _actionResult;
}
